### PR TITLE
Add Shape and Size operators

### PR DESCRIPTION
**Pull request by bddppq**
https://github.com/onnx/onnx/issues/495

### CODE REVIEW

File: docs/Changelog.md [hunk 1]
```diff
@@ -4633,6 +4633,43 @@ opset_import {
 <dd>Constrain input and output types to float tensors.</dd>
 </dl>
 
+### <a name="Shape-1"></a>**Shape-1**</a>
+
+  Takes a tensor as input and outputs an 1D int64 tensor with the shape of the input tensor.
```
Comment (jamesr66a) [inline docs/Changelog.md]: This might be unclear, maybe say `containing the shape of the input tensor` and give an example?

File: docs/Changelog.md [hunk 2]
```diff
@@ -4670,6 +4707,43 @@ opset_import {
 <dd>Constrain input and output types to float tensors.</dd>
 </dl>
 
+### <a name="Size-1"></a>**Size-1**</a>
+
+  Takes a tensor as input and outputs an 1D int64 tensor with single element that contains the total number of elements of the input tensor.
```
Comment (jamesr66a) [inline docs/Changelog.md]: Might be good to have an example here, but this one seems more clear

File: docs/Changelog.md (hunk 1 above)
Comment (bddppq) [reply docs/Changelog.md]: I don't really like to put the example in the doc string because they render poorly. The short code snippets (from the test cases) serve much better as example.

File: docs/Changelog.md (hunk 2 above)
Comment (bddppq) [reply docs/Changelog.md]: Same as above.

File: docs/Changelog.md (hunk 1 above)
Comment (bddppq) [reply docs/Changelog.md]: Wording updated

File: onnx/backend/test/case/node/size.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+class Size(Base):
+
+    @staticmethod
+    def export():
+        node = onnx.helper.make_node(
+            'Size',
+            inputs=['x'],
+            outputs=['y'],
+        )
+        x = np.random.randn(3, 4, 5).astype(np.float32)
```
Comment (houseroad) [inline onnx/backend/test/case/node/size.py]: Could you add some cases only containing manually assigned inputs/outputs, something like the case here: https://github.com/onnx/onnx/blob/master/onnx/backend/test/case/node/sqrt.py#L23

That is easier for users to understand.

File: onnx/backend/test/case/node/shape.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+class Shape(Base):
+
+    @staticmethod
+    def export():
+        node = onnx.helper.make_node(
+            'Shape',
+            inputs=['x'],
+            outputs=['y'],
+        )
+        x = np.random.randn(3, 4, 5).astype(np.float32)
```
Comment (houseroad) [inline onnx/backend/test/case/node/shape.py]: Could you add some cases only containing manually assigned inputs/outputs, something like the case here: https://github.com/onnx/onnx/blob/master/onnx/backend/test/case/node/sqrt.py#L23

That is easier for users to understand.

File: onnx/backend/test/case/node/size.py (hunk 3 above)
Comment (bddppq) [reply onnx/backend/test/case/node/size.py]: Added

File: onnx/backend/test/case/node/shape.py (hunk 4 above)
Comment (bddppq) [reply onnx/backend/test/case/node/shape.py]: Added

File: docs/Changelog.md [hunk 5]
```diff
@@ -4633,6 +4633,43 @@ opset_import {
 <dd>Constrain input and output types to float tensors.</dd>
 </dl>
 
+### <a name="Shape-1"></a>**Shape-1**</a>
+
+  Takes a tensor as input and outputs an 1D int64 tensor containing the shape of the input tensor.
```
Comment (Maratyszcza) [inline docs/Changelog.md]: Why 1D tensor I thought ONNX supports scalars as 0-dimensional tensors

**Comment by Maratyszcza** [comment]
Can we support int32/uint32 output tensors? int64 is pretty demanding, even GPUs don't natively support int64 operations, and I'm concerned accelerators may not support it at all.

File: onnx/defs/tensor/defs.cc [hunk 6]
```diff
@@ -49,6 +49,36 @@ from the shape argument.)DOC")
...
+    .TypeConstraint("T", { "tensor(float16)",  "tensor(float)", "tensor(double)",
+                           "tensor(int8)", "tensor(int16)",  "tensor(int32)", "tensor(int64)",
+                           "tensor(uint8)", "tensor(uint16)",
+                           "tensor(bool)" },
+                "Input tensor can be of arbitrary type.")
+    .TypeConstraint("T1", { "tensor(int64)"},
+	            "Constrains output to int64 tensor.");
+
+
+ONNX_OPERATOR_SCHEMA(Size)
```
[the remark this answers is not in the document]
Comment (houseroad) [reply onnx/defs/tensor/defs.cc]: Shape + ReduceSum?

[the remark this answers is not in the document]
Comment (bddppq) [reply onnx/defs/tensor/defs.cc]: @ebarsoum I thought about that, but 1) Right now onnx doesn't have Prod op (there is ReduceProd, but it's a little bit overkill to use in this case IMO) 2) we in general want to follow the numpy api and it has both `shape` and `size` (pytorch and caffe2 as well).

[the remark this answers is not in the document]
Comment (ebarsoum) [reply onnx/defs/tensor/defs.cc]: @bddppq we can add Prod OP, the problem is that how generic `size` is? For example, for batch of variable length sequences do I want to get the total number of sample per each sequence multiply by the batch size? In some cases, you would want the number of element per sample and in others you want across the batch axis. Regarding the availability of size in numpy, Caffe2 and PyTorch. This is API level not IR level, so that shouldn't impact the inclusion in the ONNX, especially if we can represented in ONNX.

How `size` in PyTorch and Caffe 2 work with variable length sequences?

[the remark this answers is not in the document]
Comment (bddppq) [reply onnx/defs/tensor/defs.cc]: @ebarsoum 
1. I agree we should add Prod, it's a fundamental math operation.
2. What do you mean by "variable" length sequences? Are you saying that each sequence could have different length?
3. IMO size is a pretty bare bone op that shouldn't take batching into consideration. numpy, caffe2, pytorch, mxnet and tensorflow return the total number.

Considering we can easily add an optimization pass right now, I don't have strong preference between size or shape + prod (although slightly prefer the former which you can tell from my arguments).

@dzhulgakov @jamesr66a thoughts?

File: docs/Changelog.md (hunk 5 above)
Comment (bddppq) [reply docs/Changelog.md]: Updated to return a scalar

File: onnx/defs/tensor/defs.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (ebarsoum) [reply onnx/defs/tensor/defs.cc]: (2) yes.
Shape and Size usually used to compute some form of normalization, Shape is generic enough and we need it.

File: docs/Changelog.md (hunk 5 above)
Comment (Maratyszcza) [reply docs/Changelog.md]: @bddppq Right, I meant the `Size` op.
